projects/components/src/overlay/overlay.service.ts

```diff
@@ -46,10 +52,13 @@ export class OverlayService implements OnDestroy {
     return popover;
   }
 
-  public createModal(config: ModalOverlayConfig, injector: Injector = this.defaultInjector): PopoverRef {
+  public createModal<TData = unknown>(
+    config: ModalOverlayConfig<TData>,
```

Comment:
I would think that we should maintain a different config for confirmation model. `ConfirmationModelConfig` may extend from ModalOverlayConfig with the addition of certain properties like `confirmButtonLabel`, `confirmClickHandler`, `cancelButtonLabel` and `cancelClickHandler`.  Based on the handler user clicks on, we can emit true/false on the `closed` observable.

Reply:
I agree with the properties, but rather than extending, I believe a ModalOverlayConfig should have a data object, and the data object is what would be injected into the modal. That would work the same way as popover. Bringing it all together:

```typescript
export interface ModalConfig <TData = never> {
  componentOrTemplate: Type<unknown> | TemplateRef<TData>;
  data?: TData; // available as MODAL_DATA
  parentInjector?: Injector;
}
```

Used by some modal, such as a confirmation modal like:
```typescript
class ConfirmationModalComponent {
  public constructor(
     private readonly modalRef: ModalRef,
     @Inject(MODAL_DATA) private readonly data: ConfirmationModalConfig) {}

   public onClose(): void {
    this.modalRef.close(true);
  }

   public onCancel(): void {
    this.modalRef.close(false);
  }
}
```